src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -0,0 +1,1388 @@
+        KeysetData keyset(fabric_index, target_id, in_keyset.policy, in_keyset.num_keys_used);
+        memcpy(keyset.epoch_keys, in_keyset.epoch_keys, sizeof(in_keyset.epoch_keys));
+        ReturnLogErrorOnFailure(keyset.Save(mStorage));
+        return fabric.Save(mStorage);
+    }
+
+    // Pseudo-index > 0
+
+    KeysetData keyset(fabric_index, fabric.first_keyset);
+    size_t keyset_count = 0;
+    bool found          = false;
+
+    // Seach for the target in_keyset id
+    while (keyset_count < fabric.keyset_count)
+    {
+        if (CHIP_NO_ERROR != keyset.Load(mStorage))
+        {
+            break;
+        }
+        if (keyset.keyset_id == target_id)
+        {
+            // Target id found
+            found = true;
+            break;
+        }
+        keyset.keyset_id = keyset.next;
+        keyset_count++;
+    }
+
+    keyset.keyset_id     = target_id;
+    keyset.policy        = in_keyset.policy;
+    keyset.num_keys_used = in_keyset.num_keys_used;
+    memcpy(keyset.epoch_keys, in_keyset.epoch_keys, sizeof(in_keyset.epoch_keys));
+
+    if (found)
+    {
+        // Update existing in_keyset info, keep next
+        return keyset.Save(mStorage);
+    }
+
+    // New in_keyset, insert first
+    keyset.next = fabric.first_keyset;
+    ReturnErrorOnFailure(keyset.Save(mStorage));
+    // Update fabric info
+    fabric.keyset_count++;
+    fabric.first_keyset = target_id;
+    return fabric.Save(mStorage);
+}
+
+CHIP_ERROR GroupDataProviderImpl::GetKeyset(chip::FabricIndex fabric_index, uint16_t target_id, Keyset & out_keyset)
+{
+    VerifyOrReturnError(mInitialized, CHIP_ERROR_INTERNAL);
+
+    FabricData fabric(fabric_index);
+    VerifyOrReturnError(CHIP_NO_ERROR == fabric.Load(mStorage), CHIP_ERROR_INVALID_FABRIC_ID);
+
+    KeysetData keyset(fabric_index, fabric.first_keyset);
+    size_t keyset_count = 0;
+
+    // Loop until the desired index
+    while (keyset_count < fabric.keyset_count)
+    {
+        if (CHIP_NO_ERROR != keyset.Load(mStorage))
+        {
+            break;
+        }
+        if (keyset.keyset_id == target_id)
+        {
+            // Target index found
+            out_keyset.keyset_id     = keyset.keyset_id;
+            out_keyset.policy        = keyset.policy;
+            out_keyset.num_keys_used = keyset.num_keys_used;
+            memcpy(out_keyset.epoch_keys, keyset.epoch_keys, sizeof(out_keyset.epoch_keys));
+            return CHIP_NO_ERROR;
+        }
+        keyset.keyset_id = keyset.next;
+        keyset_count++;
+    }
+
+    return CHIP_ERROR_KEY_NOT_FOUND;
+}
+
+CHIP_ERROR GroupDataProviderImpl::RemoveKeyset(chip::FabricIndex fabric_index, uint16_t target_id)
+{
+    VerifyOrReturnError(mInitialized, CHIP_ERROR_INTERNAL);
+
+    FabricData fabric(fabric_index);
+    VerifyOrReturnError(CHIP_NO_ERROR == fabric.Load(mStorage), CHIP_ERROR_KEY_NOT_FOUND);
+
+    KeysetData keyset(fabric_index, fabric.first_keyset);
+    uint16_t prev_id    = 0;
```

Comment:
0 means none before. I guess this is the empty node of your linked list, where 0 is not a valid index, and logical node indices are 1-based?

Reply:
Not in this case. States are stored using arbitrary but unique IDs, which may be different from the actual state index. Endpoints, groups, and keysets are stored using their actual IDs, which must be unique per fabric. So a prev_id of zero is actually valid; that's why here I use the count, not the ID, to identify the existence of previous entries.